### PR TITLE
feat: expose premium trial form via CLI

### LINKED ISSUE #14856: Expose 30d trial form via CLI
The UI requires these fields

image

However the CLI setup does not

image (1)

Thanks @dannykopping for discovering this!

### PR DESCRIPTION
**Pull request by joobisb**
This PR closes https://github.com/coder/coder/issues/14856

### CODE REVIEW

File: cli/login.go [hunk 1]
```diff
@@ -449,3 +553,75 @@ func openURL(inv *serpent.Invocation, urlToOpen string) error {
...
+		Message:    "Select the number of developers:",
+	})
+	if err != nil {
+		return "", xerrors.Errorf("select developers: %w", err)
+	}
+	return selection, nil
+}
+
+func promptCountry(inv *serpent.Invocation) (string, error) {
+	countries := []string{
```
Comment (joobisb) [inline cli/login.go:590]: this list is taken from [countries.tsx](https://github.com/coder/coder/blob/main/site/src/pages/SetupPage/countries.tsx) to have a select and filter option for this field, so the list will be maintained in two places, I'm open to options if there is a better way of handling this

File: cli/login.go [hunk 2]
```diff
@@ -149,6 +149,14 @@ func (r *RootCmd) login() *serpent.Command {
...
 		trial              bool
 		useTokenForSession bool
+
+		firstName   string
+		lastName    string
+		phoneNumber string
+		jobTitle    string
+		companyName string
+		country     string
+		developers  string
```
Comment (johnstcn) [inline cli/login.go:159]: suggestion: let's instead move these declarations closer to L278

File: cli/login.go [hunk 3]
```diff
@@ -267,12 +275,66 @@ func (r *RootCmd) login() *serpent.Command {
 					trial = v == "yes" || v == "y"
 				}
 
+				if trial {
+					if firstName == "" {
```
Comment (johnstcn) [inline cli/login.go:279]: suggestion: we can instead just leverage the zero value and assign directly to the fields of TrialInfo

e.g.

```
var trialInfo codersdk.CreateFirstUserTrialInfo
if trial {
  if trialInfo.FirstName == "" {
    ...
  }
  ...
}
```

File: cli/login.go [hunk 4]
```diff
@@ -398,6 +460,48 @@ func (r *RootCmd) login() *serpent.Command {
...
+			Env:         "CODER_FIRST_USER_COUNTRY",
+			Description: "Specifies the country of the user.",
+			Value:       serpent.StringOf(&country),
+		},
+		{
+			Flag:        "first-user-developers",
+			Env:         "CODER_FIRST_USER_DEVELOPERS",
+			Description: "Specifies the number of developers.",
+			Value:       serpent.StringOf(&developers),
+		},
```
Comment (johnstcn) [inline cli/login.go:504]: I'd rather not expose these as CLI arguments.

File: cli/login.go (hunk 1 above)
Comment (johnstcn) [reply cli/login.go:590]: I would suggest making it a separate follow-up PR, but we could instead define countries in Go and have a separate script similar to `scripts/apitypings` to generate `countries.tsx` from `countries.go`.

File: cli/login_test.go [hunk 5]
```diff
@@ -142,6 +194,12 @@ func TestLogin(t *testing.T) {
 			"password", coderdtest.FirstUserParams.Password,
 			"password", coderdtest.FirstUserParams.Password, // confirm
 			"trial", "yes",
+			"firstName", coderdtest.TrialUserParams.FirstName,
+			"lastName", coderdtest.TrialUserParams.LastName,
+			"phoneNumber", coderdtest.TrialUserParams.PhoneNumber,
+			"jobTitle", coderdtest.TrialUserParams.JobTitle,
+			"companyName", coderdtest.TrialUserParams.CompanyName,
```
Comment (johnstcn) [inline cli/login_test.go:201]: If trial is not set, then these parameters should not be required.

**Review by johnstcn — COMMENTED** [review]
Thanks for your contribution! From my perspective, the trial information should always come directly from user input. I don't feel comfortable with allowing trials to be specified directly via CLI environment variables, although I admit there's very little stopping an enterprising user from feeding information in via `expect` or similar.

File: cli/login.go (hunk 4 above)
Comment (matifali) [reply cli/login.go:504]: I agree with Cian here.

Comment (joobisb) [reply cli/login.go:504]: got it, will remove these

File: cli/login_test.go (hunk 5 above)
Comment (joobisb) [reply cli/login_test.go:201]: yeah that's right, but in this https://github.com/coder/coder/blob/10ccce1bd99210095895c27b21f04dda8d90cc30/cli/login_test.go#L196 trial is already set, so the params are required and I have added another test case `InitialUserTTYWithNoTrial` where trial is not set and params are not required
https://github.com/coder/coder/blob/10ccce1bd99210095895c27b21f04dda8d90cc30/cli/login_test.go#L128

Comment (johnstcn) [reply cli/login_test.go:201]: Oops, I must have missed that! Thanks for clarifying.

File: cli/login.go (hunk 1 above)
Comment (joobisb) [reply cli/login.go:590]: sure, will address this in a separate PR, shall I create an issue to track this ?

Comment (johnstcn) [reply cli/login.go:590]: I went ahead and created one: https://github.com/coder/coder/issues/15074

**Comment by joobisb** [comment]
> Thanks for your contribution! From my perspective, the trial information should always come directly from user input. I don't feel comfortable with allowing trials to be specified directly via CLI environment variables, although I admit there's very little stopping an enterprising user from feeding information in via `expect` or similar.

@johnstcn should we remove `first-user-trial` as well from CLI env variables ? because if its `true` then other variables would also be required, or else we should make those optional

https://github.com/coder/coder/blob/10ccce1bd99210095895c27b21f04dda8d90cc30/cli/login.go#L453

**Comment by johnstcn** [comment]
> @johnstcn should we remove `first-user-trial` as well from CLI env variables ? because if its `true` then other variables would also be required, or else we should make those optional
> 

Let's leave that as-is.

**Comment by joobisb** [comment]
> > @johnstcn should we remove `first-user-trial` as well from CLI env variables ? because if its `true` then other variables would also be required, or else we should make those optional
> 
> Let's leave that as-is.

@johnstcn 
I've addressed the comments and I have checked through the codebase, the only issue I see is that of the use of `--first-user-trial` [here](https://github.com/coder/coder/blob/fac77f956ec28ad04c0c90750e381e69465d2ae1/.github/workflows/pr-deploy.yaml#L413), since the value is set this will prompt to enter the trial details causing the workflow to fail, one option is to set it as `--first-user-trial=false`, would that be ok?

https://github.com/coder/coder/blob/fac77f956ec28ad04c0c90750e381e69465d2ae1/.github/workflows/pr-deploy.yaml#L413

**Comment by matifali** [comment]
@joobisb yes set that as false.

**Comment by joobisb** [comment]
> @joobisb yes set that as false.

i've updated the PR

File: .github/workflows/pr-deploy.yaml [hunk 6]
```diff
@@ -410,7 +410,7 @@ jobs:
             --first-user-username coder \
             --first-user-email pr${{ env.PR_NUMBER }}@coder.com \
             --first-user-password $password \
-            --first-user-trial \
+            --first-user-trial=false \
```
Comment (matifali) [inline .github/workflows/pr-deploy.yaml:413]: We are currently not making any use of a premium license in PR deployments. If needed, we can always upload one manually.

File: cli/login.go [hunk 7]
```diff
@@ -449,3 +496,75 @@ func openURL(inv *serpent.Invocation, urlToOpen string) error {
...
+		"Saint Vincent and the Grenadines", "Samoa", "San Marino", "Sao Tome and Principe", "Saudi Arabia", "Senegal", "Serbia", "Seychelles", "Sierra Leone", "Singapore",
+		"Sint Maarten (Dutch part)", "Slovakia", "Slovenia", "Solomon Islands", "Somalia", "South Africa", "South Georgia and the South Sandwich Islands", "South Sudan", "Spain", "Sri Lanka",
+		"Sudan", "Suriname", "Svalbard and Jan Mayen", "Swaziland", "Sweden", "Switzerland", "Syrian Arab Republic", "Taiwan, Province of China", "Tajikistan", "Tanzania, United Republic of",
+		"Thailand", "Timor-Leste", "Togo", "Tokelau", "Tonga", "Trinidad and Tobago", "Tunisia", "Turkey", "Turkmenistan", "Turks and Caicos Islands",
+		"Tuvalu", "Uganda", "Ukraine", "United Arab Emirates", "United Kingdom", "United States", "United States Minor Outlying Islands", "Uruguay", "Uzbekistan", "Vanuatu",
+		"Venezuela, Bolivarian Republic of", "Viet Nam", "Virgin Islands, British", "Virgin Islands, U.S.", "Wallis and Futuna", "Western Sahara", "Yemen", "Zambia", "Zimbabwe",
```
Comment (matifali) [inline cli/login.go:558]:
```suggestion
		"Venezuela, Bolivarian Republic of", "Vietnam", "Virgin Islands, British", "Virgin Islands, U.S.", "Wallis and Futuna", "Western Sahara", "Yemen", "Zambia", "Zimbabwe",
```

Comment (joobi-keyvalue) [reply cli/login.go]: this is fixed

Comment (joobisb) [reply cli/login.go]: this is fixed

**Comment by matifali** [comment]
Can you rebase on `main`?

**Review by johnstcn — COMMENTED** [review]
One thing I've noticed is that in `develop.sh` we set `--first-user-trial=true`. This breaks the `develop.sh` workflow slightly as we then prompt for trial info. How about we instead just set `--first-user-trial=false`? In a separate PR we can modify the `develop.sh` script to upload a development license from a well-known location.

**Comment by joobisb** [comment]
> One thing I've noticed is that in `develop.sh` we set `--first-user-trial=true`. This breaks the `develop.sh` workflow slightly as we then prompt for trial info. How about we instead just set `--first-user-trial=false`? In a separate PR we can modify the `develop.sh` script to upload a development license from a well-known location.

sure, I thought since it's used in the dev environment, we can always enter values into the prompt. But you are right, to make it more seamless I think we could set it to `false`.

**Comment by joobisb** [comment]
> > One thing I've noticed is that in `develop.sh` we set `--first-user-trial=true`. This breaks the `develop.sh` workflow slightly as we then prompt for trial info. How about we instead just set `--first-user-trial=false`? In a separate PR we can modify the `develop.sh` script to upload a development license from a well-known location.
> 
> sure, I thought since it's used in the dev environment, we can always enter values into the prompt. But you are right, to make it more seamless I think we could set it to `false`.

updated the PR

**Comment by joobisb** [comment]
@matifali @johnstcn can we merge this ?

**Comment by johnstcn** [comment]
@matifali it looks like the `e2e-premium` tests are failing. Can you try rebasing your PR on latest `main`?

**Comment by joobi-keyvalue** [comment]
> @matifali it looks like the `e2e-premium` tests are failing. @joobisb Can you try rebasing your PR on latest `main`?

@johnstcn done

**Comment by joobisb** [comment]
> @matifali it looks like the `e2e-premium` tests are failing. @joobisb Can you try rebasing your PR on latest `main`?

@johnstcn done

**Comment by joobisb** [comment]
@matifali can we merge this ?

**Review by matifali — APPROVED** [review]

**Comment by matifali** [comment]
@sreya, we are blocked here as the secret is unavailable for forks. This is a required test so can not merge without it being passed.

**Pull request merged**